ModuleFast fails in an Azure DevOps Release Pipeline

### Problem description

Building a project with ~80 dependencies fail with weird error messages when using ModuleFast. The errors are like:

```
[pre-build] Starting bootstrap process.
Both ModuleFast and PSResourceGet is configured or/and passed as parameter.
PowerShell 7.2 or higher being used, prefer ModuleFast over PSResourceGet.
ModuleFast is configured to use latest released version.
Install-ModuleFast : Cannot convert value to type System.String.
At C:\Agent\_work\1\s\Resolve-Dependency.ps1:871 char:39
+ …     $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidResult: (:) [Install-ModuleFast], RuntimeException
+ FullyQualifiedErrorId : JobStateFailed,Install-ModuleFast
##[error]PowerShell exited with code '1'.
```

or

```
Install-ModuleFast : Exception calling ".ctor" with "2" argument(s): "The HTTP/3 server reset the stream. HTTP/3 error code 'H3_NO_ERROR' (0x100). (HttpProtocolError)"
At C:\Agent\_work\1\s\Resolve-Dependency.ps1:871 char:39
+ …     $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidResult: (:) [Install-ModuleFast], MethodInvocationException
+ FullyQualifiedErrorId : JobStateFailed,Install-ModuleFast
##[error]PowerShell exited with code '1'.
```



### Verbose logs

```text
NA
```

### How to reproduce

Build [Microsoft365DscWorkshop](https://github.com/raandree/Microsoft365DscWorkshop/blob/feature/labUpdate/docs/GettingStarted.md) in an Azure pipeline.

### Expected behavior

Some error handling should be added. I have noticed that the issue disappears almost completely after retrying.

### Current behavior

Ignores errors in ModuleFast and continues without having all the dependencies downloaded.

### Suggested solution

Add error handling / retry behaviour.

### Operating system the target node is running

```text
Azure Pipelines
```

### PowerShell version and build the target node is running

```text
PowerShell 7 on Azure Pipelines
```

### Module version used

```text
0.118.2
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- Add retry logic for ModuleFast dependency installation.
+
 ## [0.118.3] - 2025-04-29
 
 ### Changed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
-- Add retry logic for ModuleFast dependency installation.
+- Add retry logic for ModuleFast dependency installation ([#510](https://github.com/gaelcolas/Sampler/issues/510)).
 
 ## [0.118.3] - 2025-04-29
 

--- a/Resolve-Dependency.ps1
+++ b/Resolve-Dependency.ps1
@@ -866,9 +866,24 @@ try
                 if ($moduleFastPlan)
                 {
                     # Clear all modules in plan from the current session so they can be fetched again.
-                    $moduleFastPlan.Name | Get-Module | Remove-Module -Force
-
-                    $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                    try
+                    {
+                        $moduleFastPlan.Name | Get-Module | Remove-Module -Force
+                        $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                    }
+                    catch
+                    {
+                        Write-Warning -Message 'ModuleFast could not save one or more dependencies. Retrying...'
+                        try
+                        {
+                            $moduleFastPlan.Name | Get-Module | Remove-Module -Force
+                            $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                        }
+                        catch
+                        {
+                            Write-Error 'ModuleFast could not save one or more dependencies even after a retry.'
+                        }
+                    }
                 }
                 else
                 {

--- a/Sampler/Templates/Build/Resolve-Dependency.ps1
+++ b/Sampler/Templates/Build/Resolve-Dependency.ps1
@@ -866,9 +866,24 @@ try
                 if ($moduleFastPlan)
                 {
                     # Clear all modules in plan from the current session so they can be fetched again.
-                    $moduleFastPlan.Name | Get-Module | Remove-Module -Force
-
-                    $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                    try
+                    {
+                        $moduleFastPlan.Name | Get-Module | Remove-Module -Force
+                        $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                    }
+                    catch
+                    {
+                        Write-Warning -Message 'ModuleFast could not save one or more dependencies. Retrying...'
+                        try
+                        {
+                            $moduleFastPlan.Name | Get-Module | Remove-Module -Force
+                            $moduleFastPlan | Install-ModuleFast @installModuleFastParameters
+                        }
+                        catch
+                        {
+                            Write-Error 'ModuleFast could not save one or more dependencies even after a retry.'
+                        }
+                    }
                 }
                 else
                 {

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -35,6 +35,7 @@ stages:
               }
               Write-Host -Object "##vso[build.updatebuildnumber]$($gitVersionObject.FullSemVer)"
             displayName: Calculate ModuleVersion (GitVersion)
+
           - task: PowerShell@2
             name: package
             displayName: 'Build & Package Module'

